file.go

```diff
@@ -210,20 +222,78 @@ func (f *win32File) Write(b []byte) (int, error) {
 	if err != nil {
 		return 0, err
 	}
+
+	timeout := f.writeTimeoutChan
+	if f.writeTimedOut.isSet() {
+		return 0, ErrTimeout
+	}
+
 	var bytes uint32
 	err = syscall.WriteFile(f.handle, b, &bytes, &c.o)
-	n, err := f.asyncIo(c, f.writeDeadline, bytes, err)
+	n, err := f.asyncIo(c, timeout, bytes, err)
 	runtime.KeepAlive(b)
 	return n, err
 }
 
-func (f *win32File) SetReadDeadline(t time.Time) error {
-	f.readDeadline = t
+func (f *win32File) SetReadDeadline(deadline time.Time) error {
+	now := time.Now()
+	duration := deadline.Sub(now)
+
+	if f.readTimer != nil {
+		f.readTimer.Stop()
+	}
+	f.readTimer = nil
+	f.readTimedOut.setFalse()
+
+	if deadline.IsZero() {
+		return nil
+	}
+
+	timeoutReads := func() {
+		f.readTimedOut.setTrue()
+		close(f.readTimeoutChan)
+		// TODO sync here???
+		f.readTimeoutChan = make(chan struct{})
+	}
+
+	if deadline.After(now) {
+		// Deadline is in the future, set a timer to wait
+		f.readTimer = time.AfterFunc(duration, timeoutReads)
+	} else {
+		// Deadline is in the past. Cancel all pending reads now.
+		timeoutReads()
+	}
 	return nil
 }
 
-func (f *win32File) SetWriteDeadline(t time.Time) error {
-	f.writeDeadline = t
+func (f *win32File) SetWriteDeadline(deadline time.Time) error {
+	now := time.Now()
+	duration := deadline.Sub(now)
+
+	if f.writeTimer != nil {
+		f.writeTimer.Stop()
+	}
+	f.writeTimer = nil
+	f.writeTimedOut.setFalse()
+
+	if deadline.IsZero() {
+		return nil
+	}
+
+	timeoutWrites := func() {
+		f.writeTimedOut.setTrue()
+		close(f.writeTimeoutChan)
+		// TODO sync here???
+		f.writeTimeoutChan = make(chan struct{})
```

Comment:
I don't understand. If a new read or write comes in, it would wait on the closed channel which should complete immediately. That's what we want, right?